### ISSUE TITLE
Reassess names, fields, and groups for CustomResources

### ISSUE THREAD

**Issue by deads2k** [kubernetes/kubernetes#45277]
https://github.com/kubernetes/kubernetes/pull/45115 lays out initial types for `CustomResources` (née TPR).  The API wasn't exposed anywhere (only testable via an aggregated API server), so to make progress to the meat of the CustomResources implementation, we merged it fairly quickly.

This issue is a spot to talk about what we don't like about the names and what we'd like changed.  Pulls are also welcome, but we can discuss it here without all the trouble of writing code.

**Comment by deads2k** [comment]
Types are located here: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/pkg/apis/extensions/v1beta1/types.go .

**Comment by deads2k** [comment]
Guess I'll start.  I think we need clarity between the `CustomResourceDefinition` and the "custom resource" that is provided.  I think including `Definition` in the name might look weird, but it will clarify any future discussion and documentation.

**Comment by enisoc** [comment]
CC @foxish who has expressed opinions on this previously.

TL;DR: I'm in favor of the names @deads2k has proposed. Unless there are objections, I say we move forward.

First of all, I like the base name `CustomResource` a lot:

* "Custom" implies the resource is not necessarily available on all servers (it's a customization).
  * The term "third party" didn't get at this directly, and was potentially misleading. The party that provided the resource definition is not really the important dimension to emphasize. What if we offer first-party customizations? What if the users themselves create customizations (technically second-party)?
* It retains the word `Resource`, which I think is valuable because the places where things like `kubectl` use that word are exactly where you expect these custom things to show up as well.
* It ties in well with discussion of "custom controllers".
* It flows better verbally.

Second, to add context for the proposal of `CustomResourceDefinition`: the problem we saw with TPR is that people would use the term TPR to refer to both:
1. objects of kind `ThirdPartyResource` that define a new type of resource,

   and

2. instances of that new resource type (e.g. objects of kind `EtcdCluster`).

That made it difficult to unambiguously refer to one or the other. Users still may be tempted to use CRD improperly, but at least we'd have the vocabulary to document the difference between CR and CRD, which we didn't have with the name TPR.
